operator/pkg/clustermesh/pod_informer.go

```diff
@@ -0,0 +1,218 @@
+type meshPodInformer struct {
+	GlobalServiceCache *globalServiceCache
+	handler            cache.ResourceEventHandler
+}
+
+func getContainerPorts(portConfig store.PortConfiguration) []v1.ContainerPort {
+	ports := []v1.ContainerPort{}
+	for name, l4Addr := range portConfig {
+		ports = append(ports, v1.ContainerPort{
+			Name:          name,
+			ContainerPort: int32(l4Addr.Port),
+			Protocol:      v1.Protocol(l4Addr.Protocol),
+		})
+	}
+	return ports
+}
+
+// TODO: handle pod hostname
+func podFromSingleAddr(addr string, portConfig store.PortConfiguration, clusterSvc *store.ClusterService) *v1.Pod {
+	return &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "cilium.io",
+			Kind:       "store.ClusterService",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-pod-" + addr,
+			Namespace: clusterSvc.Namespace,
+			Labels: map[string]string{
+				meshServiceKeyLabel:     clusterSvc.NamespaceServiceName(),
+				meshServiceClusterLabel: clusterSvc.Cluster,
+			},
+		},
+		Spec: v1.PodSpec{
+			NodeName: clusterSvc.Cluster,
+			Containers: []v1.Container{{
+				Ports: getContainerPorts(portConfig),
+			}},
+		},
+		Status: v1.PodStatus{
+			Phase: v1.PodRunning,
+			Conditions: []v1.PodCondition{
+				{Type: v1.PodReady, Status: v1.ConditionTrue},
+			},
+			PodIPs: []v1.PodIP{{IP: addr}},
+		},
+	}
+}
+
+func validateSelector(reqs labels.Requirements) (string, string, error) {
+	var key *string
+	var cluster *string
+
+	if len(reqs) != 2 {
+		goto err
+	}
+
+	for _, req := range reqs {
+		if req.Values().Len() != 1 || req.Operator() != selection.Equals {
+			goto err
+		}
+
+		if req.Key() == meshServiceKeyLabel {
+			if key != nil {
+				goto err
+			}
+			key = &req.Values().List()[0]
+		} else if req.Key() == meshServiceClusterLabel {
+			if cluster != nil {
+				goto err
+			}
+			cluster = &req.Values().List()[0]
+		}
+	}
+
+	if key == nil || cluster == nil {
+		goto err
+	}
+
+	return *key, *cluster, nil
+
+err:
+	return "", "", fmt.Errorf("meshPodInformer only supports listing with both service name and cluster as requirements: %s", reqs)
+}
+
+func (i meshPodInformer) List(selector labels.Selector) ([]*v1.Pod, error) {
+	reqs, _ := selector.Requirements()
+	key, cluster, err := validateSelector(reqs)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterSvc := i.GlobalServiceCache.getService(key, cluster)
+	if clusterSvc == nil {
+		return nil, nil
+	}
+
+	pods := make([]*v1.Pod, 0, len(clusterSvc.Backends))
+	for addr, portConfig := range clusterSvc.Backends {
+		pods = append(pods, podFromSingleAddr(addr, portConfig, clusterSvc))
+	}
+	return pods, nil
+}
+
+func (i *meshPodInformer) AddEventHandler(handler cache.ResourceEventHandler) (cache.ResourceEventHandlerRegistration, error) {
+	i.handler = handler
+	return i, nil
+}
+
+func (i *meshPodInformer) onClusterServiceUpdate(clusterSvc *store.ClusterService) {
+	if i.handler == nil {
+		return
+	}
+	if len(clusterSvc.Backends) == 0 {
```

Comment:
Hmmm well the onDelete hook in the controller behind the scene is pretty much the same as `addPod` (https://github.com/cilium/endpointslice-controller/blob/main/endpointslice/endpointslice_controller.go#L554), IIRC this code was only added because in this case I didn't have a backend to generate a pod from (calling `podFromSingleAdd`).

Technically I guess I could always call `onAdd` with a dummy object without calling `podFromSingleAddr` so this doesn't have two code path, WDYT?

Reply:
> Hmmm well the onDelete hook in the controller behind the scene is pretty much the same as addPod 

Fair enough

> Technically I guess I could always call onAdd with a dummy object without calling podFromSingleAddr so this doesn't have two code path, WDYT?

Sounds reasonable to me. I'd still add a comment to clarify that, especially because also the `Name` and `Labels` look weird (as in different from the ones configured by `podFromSingleAddr`).

A follow-up comment is that it seems we also don't generate a Delete (or Add fwiw) event for the backends that get removed from a cluster service, which will likely cause them not to be removed from the epslices.

Reply:
> Sounds reasonable to me. I'd still add a comment to clarify that, especially because also the Name and Labels look weird (as in different from the ones configured by podFromSingleAddr).

:+1: 

> A follow-up comment is that it seems we also don't generate a Delete (or Add fwiw) event for the backends that get removed from a cluster service, which will likely cause them not to be removed from the epslices.

The remoteServiceObserver `onDelete` also calls this function so I think the logic is fine but that's confusing indeed, I will probably do two functions with one calling the other to make this more explicit.

Reply:
Right, I always forget that it then indirectly triggers the enqueue of the entire service. I'd suggest adding a comment to clarify the intent. And it looks like the labels should also include the name of the service in addition to that of the cluster, otherwise they are not matched by the [service selector](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/operator/pkg/clustermesh/service_informer.go#L204-L205). 

Another possibility could be directly triggering a service update/delete event for that cluster on global service changes through the service informer, instead of going through the pod informer. That would be likely clearer, but my understanding is that we would then enqueue it immediately, rather than after the batch period. To me, it looks like it should not make any difference in practice, though, given that global services are already constructed out of endpointslices in the remote cluster (unless epslices support is disabled, but I guess that's never the case), and multiple updates are also [potentially coalescenced by the synchronization logic](https://github.com/MrFreezeex/cilium/blob/add-sync-endpoints/operator/watchers/k8s_service_sync.go#L58-L68).

Reply:
Oh that make sense so indeed batching at this point might be not very useful :+1:.

The only thing is that I'm wondering if the rate limiting would be sufficient in really busy cluster and if we need to do this extra batching, but I guess we might go for the simpler thing here (no batching) and determine at a later point if we need to re-introduce this or not.

Reply:
> The only thing is that I'm wondering if the kvstore rate limiting would be sufficient in a really busy cluster

Consider that every new pod that gets created triggers also an upsertion of new key in etcd (independently of the services logic), and that the default rate limiter is 20 QPS (it could be raised, but still not to thousands of QPS). So I'd expect global service updates to be always batched even more in that case.

Still, I'd prefer to also hear Marcel's opinion on this.

Reply:
> It would be slightly different if the service reconciliation was keyed by ns/name instead that by cluster/ns/name, as in that case we could receive multiple updates for the same global service from different remote clusters at the same time. Still probably unlikely in most scenarios, but definitely possible in large clustermeshes for services including backends in multiple clusters.

ah yea, when I was thinking about batching I thought it will be `ns/name` and not `cluster/ns/name`.  In that case it's fine not to batch I guess, although the default in kube-controller-manager is without batching, so we potentially might see 1 update of endpointslice per pod in remote cluster (unless clustermesh batches them because of 20 qps in remote cluster) so it could be still useful - imagine remote cluster without batching in kcm, we would have 10 qps for pod updates and 10 qps for service updates in remote cluster.

Reply:
> although the default in kube-controller-manager is without batching

Ah, I thought batching was enabled by default.

Reply:
Just to clarify do we want to do no batching by default but keep the current behavior of faking pod to notify of remote cluster service updates which means that users could tweak this optionally and introduce batching? Or do we simplify the code a bit and notify the controller via service events meaning no batching possible at all.

Reply:
I'm not sure that we already reached full consensus.

My understanding is that there are three main possibilities:
* Notify through fake pods - Pros: batching, Cons: the logic is more complex and less intuitive (in my personal opinion), because we receive a global service update which is then faked as a pod update for a single backend, relying on the fact that under the hood the epslice controller then enqueues the service.
* Notify through fake services - Pros: the logic seems simpler and more straightforward to me. Cons: we would not have batching.
* Notify through fake services, and amend the upstream fork with `AddAfter()` instead of `Add()` for services. Pros: both the ones above. Cons: one more change in that fork.

I don't have personally too strong opinions on any of the approaches. It also sounds reasonable to me to just go for the first, as already implemented. Though, I'd suggest creating a separate function (instead of `podFromSingleAddr`) which just initializes the ObjectMeta with a name matching the service, instead of one of the backends. That way, it would be the same regardless of whether it has backends or not. And add a comment to clarify the reasoning, so that it still makes sense when we look again at the same code in some weeks/months.